src/Client/GetHedgedConnections.h

```diff
@@ -0,0 +1,173 @@
+#pragma once
+
+#if defined(OS_LINUX)
+
+#include <Common/TimerDescriptor.h>
+#include <Client/ConnectionPoolWithFailover.h>
+#include <Core/Settings.h>
+#include <Common/Epoll.h>
+#include <unordered_map>
+#include <memory>
+
+namespace DB
+{
+
+using TimerDescriptorPtr = std::shared_ptr<TimerDescriptor>;
+
+/// Class for establishing hedged connections with replicas.
+/// It works with multiple replicas simultaneously without blocking by using epoll.
+class GetHedgedConnections
+{
+public:
+    using ShuffledPool = ConnectionPoolWithFailover::Base::ShuffledPool;
+
+    enum State
+    {
+        EMPTY = 0,
+        READY = 1,
+        NOT_READY = 2,
+        CANNOT_CHOOSE = 3,
+    };
+
+    struct ReplicaState
```

Comment:
This struct is very general. Some parts of it is used only in `HedgedConnections`, other parts only in `GetHedgedConnections`. 

I suppose, all methods which get connections should only return `State` (`EMPTY` should be removed, cause it is not used outside) and `Connection *` (if state is `READY`)